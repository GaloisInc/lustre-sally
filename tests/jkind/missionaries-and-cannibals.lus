-- A Lustre encoding of the "Missionaries and cannibals problem"
-- https://en.wikipedia.org/wiki/Missionaries_and_cannibals_problem
-- Contributed by Steven Miller
--
-- This file has 3 true properties at K = 1 and one false at K = 12

type bank = struct { m: int; c:int };

type state = struct { left: bank; right: bank };

const missionaries = 3;
const cannibals = 3;

const initial_state = state { left  = bank {m = missionaries; c = cannibals};
                              right = bank {m = 0; c = 0}};

-- constraint on cannibals and missionaries transported from source to destination bank
node safe(src, dst : bank; m, c : int) returns (r: bool);
let
  r = m >= 0 and c >= 0 and                      -- non-negative occupants in boat
      m + c >= 1 and m + c <= 2 and              -- 1 to 2 occupants in boat
      m <= src.m and c <= src.c and              -- occupants on source bank
      (src.m - m >= src.c - c or src.m = m) and  -- source bank will be safe
      (dst.m + m >= dst.c + c or dst.m + m = 0); -- destination bank will be safe
tel;

node ferry_right(s:state; m, c : int) returns (r : state);
let
  r = state { left = bank {m = s.left.m - m; c = s.left.c - c};
              right = bank {m = s.right.m + m; c = s.right.c + c}};
tel;

node ferry_left(s:state; m, c : int) returns (r : state);
let
  r = state { left = bank {m = s.left.m + m; c = s.left.c + c};
              right = bank {m = s.right.m - m; c = s.right.c - c}};
tel;

node main(m, c : int) returns ();
var
  s : state;
  boat_on_left_bank : bool;

  -- properties to prove
  conservation_missionaries, conservation_cannibals, safety_missionaries, cex : bool;
let
  -- boat alternates from left to right bank
  boat_on_left_bank = (true -> not pre boat_on_left_bank);

  -- constrain missionaries and cannibals in boat to maintain missionary safety
  assert if (boat_on_left_bank)
         then safe(s.left, s.right, m, c)
         else safe(s.right, s.left, m, c);

  -- compute next states
  s = initial_state ->
        if (pre boat_on_left_bank) then
           ferry_right(pre s, pre m, pre c)
        else
           ferry_left(pre s, pre m, pre c);

  conservation_missionaries = (s.left.m + s.right.m = missionaries);
  --%PROPERTY conservation_missionaries;

  conservation_cannibals = (s.left.c + s.right.c = cannibals);
  --%PROPERTY conservation_cannibals;

  safety_missionaries = ((s.left.m >= s.left.c or s.left.m = 0) and
                         (s.right.m >= s.right.c or s.right.m = 0));
  --%PROPERTY safety_missionaries;

  cex = not (s.right.m = missionaries and s.right.c = cannibals);
  --%PROPERTY cex;
tel;
