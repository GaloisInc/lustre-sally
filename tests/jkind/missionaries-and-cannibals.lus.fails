TODO: See #1 on Github
