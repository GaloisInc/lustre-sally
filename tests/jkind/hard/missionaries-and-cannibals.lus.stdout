Core Lustre
===========

node (m_22 : int, c_23 : int)
  returns ()
  assumes (ITE)
  shows (cex, safety_missionaries, conservation_cannibals, conservation_missionaries)
let
  valof_pre_boat_on_left_bank_90 = pre boat_on_left_bank
  boat_on_left_bank = true -> Not(valof_pre_boat_on_left_bank_90)
  s.left.m_84 = pre s.left.m_62
  m_88 = pre m_22
  __no_static.left.m_54 = Add(s.left.m_84, m_88)
  s.left.m_78 = pre s.left.m_62
  m_82 = pre m_22
  __no_static.left.m_58 = Sub(s.left.m_78, m_82)
  valof_pre_boat_on_left_bank_92 = pre boat_on_left_bank
  s.left.m_62 = 3 -> ITE(valof_pre_boat_on_left_bank_92, __no_static.left.m_58, __no_static.left.m_54)
  s.right.m_86 = pre s.right.m_64
  __no_static.right.m_56 = Sub(s.right.m_86, m_88)
  s.right.m_80 = pre s.right.m_64
  __no_static.right.m_60 = Add(s.right.m_80, m_82)
  valof_pre_boat_on_left_bank_94 = pre boat_on_left_bank
  s.right.m_64 = 0 -> ITE(valof_pre_boat_on_left_bank_94, __no_static.right.m_60, __no_static.right.m_56)
  conservation_missionaries = Eq(Add(s.left.m_62, s.right.m_64), 3)
  s.left.c_85 = pre s.left.c_63
  c_89 = pre c_23
  __no_static.left.c_55 = Add(s.left.c_85, c_89)
  s.left.c_79 = pre s.left.c_63
  c_83 = pre c_23
  __no_static.left.c_59 = Sub(s.left.c_79, c_83)
  valof_pre_boat_on_left_bank_93 = pre boat_on_left_bank
  s.left.c_63 = 3 -> ITE(valof_pre_boat_on_left_bank_93, __no_static.left.c_59, __no_static.left.c_55)
  s.right.c_87 = pre s.right.c_65
  __no_static.right.c_57 = Sub(s.right.c_87, c_89)
  s.right.c_81 = pre s.right.c_65
  __no_static.right.c_61 = Add(s.right.c_81, c_83)
  valof_pre_boat_on_left_bank_95 = pre boat_on_left_bank
  s.right.c_65 = 0 -> ITE(valof_pre_boat_on_left_bank_95, __no_static.right.c_61, __no_static.right.c_57)
  conservation_cannibals = Eq(Add(s.left.c_63, s.right.c_65), 3)
  safety_missionaries = And(Or(Geq(s.left.m_62, s.left.c_63), Eq(s.left.m_62, 0)), Or(Geq(s.right.m_64, s.right.c_65), Eq(s.right.m_64, 0)))
  cex = Not(And(Eq(s.right.m_64, 3), Eq(s.right.c_65, 3)))
  src.m_66 = s.left.m_62
  src.c_67 = s.left.c_63
  dst.m_68 = s.right.m_64
  dst.c_69 = s.right.c_65
  m_70 = m_22
  c_71 = c_23
  __no_static_30 = And(And(And(And(And(And(And(Geq(m_70, 0), Geq(c_71, 0)), Geq(Add(m_70, c_71), 1)), Leq(Add(m_70, c_71), 2)), Leq(m_70, src.m_66)), Leq(c_71, src.c_67)), Or(Geq(Sub(src.m_66, m_70), Sub(src.c_67, c_71)), Eq(src.m_66, m_70))), Or(Geq(Add(dst.m_68, m_70), Add(dst.c_69, c_71)), Eq(Add(dst.m_68, m_70), 0)))
  src.m_72 = s.right.m_64
  src.c_73 = s.right.c_65
  dst.m_74 = s.left.m_62
  dst.c_75 = s.left.c_63
  m_76 = m_22
  c_77 = c_23
  __no_static_31 = And(And(And(And(And(And(And(Geq(m_76, 0), Geq(c_77, 0)), Geq(Add(m_76, c_77), 1)), Leq(Add(m_76, c_77), 2)), Leq(m_76, src.m_72)), Leq(c_77, src.c_73)), Or(Geq(Sub(src.m_72, m_76), Sub(src.c_73, c_77)), Eq(src.m_72, m_76))), Or(Geq(Add(dst.m_74, m_76), Add(dst.c_75, c_77)), Eq(Add(dst.m_74, m_76), 0)))
  ITE = ITE(boat_on_left_bank, __no_static_30, __no_static_31)
tel
Sally Model
===========

(define-state-type
   S
   (
     (ITE Bool)
     (cex Bool)
     (|__no_static.left.c:55| Int)
     (|__no_static.left.c:59| Int)
     (|__no_static.left.m:54| Int)
     (|__no_static.left.m:58| Int)
     (|__no_static.right.c:57| Int)
     (|__no_static.right.c:61| Int)
     (|__no_static.right.m:56| Int)
     (|__no_static.right.m:60| Int)
     (|__no_static:30| Bool)
     (|__no_static:31| Bool)
     (|boat_on_left_bank:init| Bool)
     (|boat_on_left_bank| Bool)
     (|c:23| Int)
     (|c:71| Int)
     (|c:77| Int)
     (|c:83| Int)
     (|c:89| Int)
     (|conservation_cannibals| Bool)
     (|conservation_missionaries| Bool)
     (|dst.c:69| Int)
     (|dst.c:75| Int)
     (|dst.m:68| Int)
     (|dst.m:74| Int)
     (|gal-initializing| Bool)
     (|m:22| Int)
     (|m:70| Int)
     (|m:76| Int)
     (|m:82| Int)
     (|m:88| Int)
     (|s.left.c:63:init| Bool)
     (|s.left.c:63| Int)
     (|s.left.c:79| Int)
     (|s.left.c:85| Int)
     (|s.left.m:62:init| Bool)
     (|s.left.m:62| Int)
     (|s.left.m:78| Int)
     (|s.left.m:84| Int)
     (|s.right.c:65:init| Bool)
     (|s.right.c:65| Int)
     (|s.right.c:81| Int)
     (|s.right.c:87| Int)
     (|s.right.m:64:init| Bool)
     (|s.right.m:64| Int)
     (|s.right.m:80| Int)
     (|s.right.m:86| Int)
     (|safety_missionaries| Bool)
     (|src.c:67| Int)
     (|src.c:73| Int)
     (|src.m:66| Int)
     (|src.m:72| Int)
     (|valof_pre_boat_on_left_bank:90| Bool)
     (|valof_pre_boat_on_left_bank:92| Bool)
     (|valof_pre_boat_on_left_bank:93| Bool)
     (|valof_pre_boat_on_left_bank:94| Bool)
     (|valof_pre_boat_on_left_bank:95| Bool))
   (
     (|c:23| Int)
     (|m:22| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |boat_on_left_bank:init| true)
      (= |s.left.m:62:init| true)
      (= |s.right.m:64:init| true)
      (= |s.left.c:63:init| true)
      (= |s.right.c:65:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|m:22| input.|m:22|)
      (= next.|c:23| input.|c:23|)
      next.ITE
      (= next.|valof_pre_boat_on_left_bank:90| state.|boat_on_left_bank|)
      (ite
         state.|boat_on_left_bank:init|
         (= next.|boat_on_left_bank| true)
         (=
            next.|boat_on_left_bank|
            (not next.|valof_pre_boat_on_left_bank:90|)))
      (= next.|boat_on_left_bank:init| false)
      (= next.|s.left.m:84| state.|s.left.m:62|)
      (= next.|m:88| state.|m:22|)
      (=
         next.|__no_static.left.m:54|
         (+ next.|s.left.m:84| next.|m:88|))
      (= next.|s.left.m:78| state.|s.left.m:62|)
      (= next.|m:82| state.|m:22|)
      (=
         next.|__no_static.left.m:58|
         (- next.|s.left.m:78| next.|m:82|))
      (= next.|valof_pre_boat_on_left_bank:92| state.|boat_on_left_bank|)
      (ite
         state.|s.left.m:62:init|
         (= next.|s.left.m:62| 3)
         (=
            next.|s.left.m:62|
            (ite next.|valof_pre_boat_on_left_bank:92| next.|__no_static.left.m:58| next.|__no_static.left.m:54|)))
      (= next.|s.left.m:62:init| false)
      (= next.|s.right.m:86| state.|s.right.m:64|)
      (=
         next.|__no_static.right.m:56|
         (- next.|s.right.m:86| next.|m:88|))
      (= next.|s.right.m:80| state.|s.right.m:64|)
      (=
         next.|__no_static.right.m:60|
         (+ next.|s.right.m:80| next.|m:82|))
      (= next.|valof_pre_boat_on_left_bank:94| state.|boat_on_left_bank|)
      (ite
         state.|s.right.m:64:init|
         (= next.|s.right.m:64| 0)
         (=
            next.|s.right.m:64|
            (ite next.|valof_pre_boat_on_left_bank:94| next.|__no_static.right.m:60| next.|__no_static.right.m:56|)))
      (= next.|s.right.m:64:init| false)
      (=
         next.|conservation_missionaries|
         (=
            (+ next.|s.left.m:62| next.|s.right.m:64|)
            3))
      (= next.|s.left.c:85| state.|s.left.c:63|)
      (= next.|c:89| state.|c:23|)
      (=
         next.|__no_static.left.c:55|
         (+ next.|s.left.c:85| next.|c:89|))
      (= next.|s.left.c:79| state.|s.left.c:63|)
      (= next.|c:83| state.|c:23|)
      (=
         next.|__no_static.left.c:59|
         (- next.|s.left.c:79| next.|c:83|))
      (= next.|valof_pre_boat_on_left_bank:93| state.|boat_on_left_bank|)
      (ite
         state.|s.left.c:63:init|
         (= next.|s.left.c:63| 3)
         (=
            next.|s.left.c:63|
            (ite next.|valof_pre_boat_on_left_bank:93| next.|__no_static.left.c:59| next.|__no_static.left.c:55|)))
      (= next.|s.left.c:63:init| false)
      (= next.|s.right.c:87| state.|s.right.c:65|)
      (=
         next.|__no_static.right.c:57|
         (- next.|s.right.c:87| next.|c:89|))
      (= next.|s.right.c:81| state.|s.right.c:65|)
      (=
         next.|__no_static.right.c:61|
         (+ next.|s.right.c:81| next.|c:83|))
      (= next.|valof_pre_boat_on_left_bank:95| state.|boat_on_left_bank|)
      (ite
         state.|s.right.c:65:init|
         (= next.|s.right.c:65| 0)
         (=
            next.|s.right.c:65|
            (ite next.|valof_pre_boat_on_left_bank:95| next.|__no_static.right.c:61| next.|__no_static.right.c:57|)))
      (= next.|s.right.c:65:init| false)
      (=
         next.|conservation_cannibals|
         (=
            (+ next.|s.left.c:63| next.|s.right.c:65|)
            3))
      (=
         next.|safety_missionaries|
         (and
            (or
               (<= next.|s.left.c:63| next.|s.left.m:62|)
               (= next.|s.left.m:62| 0))
            (or
               (<= next.|s.right.c:65| next.|s.right.m:64|)
               (= next.|s.right.m:64| 0))))
      (=
         next.cex
         (not
            (and
               (= next.|s.right.m:64| 3)
               (= next.|s.right.c:65| 3))))
      (= next.|src.m:66| next.|s.left.m:62|)
      (= next.|src.c:67| next.|s.left.c:63|)
      (= next.|dst.m:68| next.|s.right.m:64|)
      (= next.|dst.c:69| next.|s.right.c:65|)
      (= next.|m:70| next.|m:22|)
      (= next.|c:71| next.|c:23|)
      (=
         next.|__no_static:30|
         (and
            (<= 0 next.|m:70|)
            (<= 0 next.|c:71|)
            (<=
               1
               (+ next.|m:70| next.|c:71|))
            (<=
               (+ next.|m:70| next.|c:71|)
               2)
            (<= next.|m:70| next.|src.m:66|)
            (<= next.|c:71| next.|src.c:67|)
            (or
               (<=
                  (- next.|src.c:67| next.|c:71|)
                  (- next.|src.m:66| next.|m:70|))
               (= next.|src.m:66| next.|m:70|))
            (or
               (<=
                  (+ next.|dst.c:69| next.|c:71|)
                  (+ next.|dst.m:68| next.|m:70|))
               (=
                  (+ next.|dst.m:68| next.|m:70|)
                  0))))
      (= next.|src.m:72| next.|s.right.m:64|)
      (= next.|src.c:73| next.|s.right.c:65|)
      (= next.|dst.m:74| next.|s.left.m:62|)
      (= next.|dst.c:75| next.|s.left.c:63|)
      (= next.|m:76| next.|m:22|)
      (= next.|c:77| next.|c:23|)
      (=
         next.|__no_static:31|
         (and
            (<= 0 next.|m:76|)
            (<= 0 next.|c:77|)
            (<=
               1
               (+ next.|m:76| next.|c:77|))
            (<=
               (+ next.|m:76| next.|c:77|)
               2)
            (<= next.|m:76| next.|src.m:72|)
            (<= next.|c:77| next.|src.c:73|)
            (or
               (<=
                  (- next.|src.c:73| next.|c:77|)
                  (- next.|src.m:72| next.|m:76|))
               (= next.|src.m:72| next.|m:76|))
            (or
               (<=
                  (+ next.|dst.c:75| next.|c:77|)
                  (+ next.|dst.m:74| next.|m:76|))
               (=
                  (+ next.|dst.m:74| next.|m:76|)
                  0))))
      (=
         next.ITE
         (ite next.|boat_on_left_bank| next.|__no_static:30| next.|__no_static:31|))))

(query
   TS
   (or |gal-initializing| cex))
(query
   TS
   (or |gal-initializing| |safety_missionaries|))
(query
   TS
   (or |gal-initializing| |conservation_cannibals|))
(query
   TS
   (or |gal-initializing| |conservation_missionaries|))

[Lustre] Validating properties:
[Lustre] Property cex... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
considering simultaneous states to depth 2
considering simultaneous states to depth 2
considering simultaneous states to depth 3
considering simultaneous states to depth 3
considering simultaneous states to depth 4
considering simultaneous states to depth 4
considering simultaneous states to depth 5
considering simultaneous states to depth 5
considering simultaneous states to depth 6
considering simultaneous states to depth 6
considering simultaneous states to depth 7
considering simultaneous states to depth 7
considering simultaneous states to depth 8
considering simultaneous states to depth 8
considering simultaneous states to depth 9
considering simultaneous states to depth 9
considering simultaneous states to depth 10
considering simultaneous states to depth 10
counter-example search depth 0
counter-example search depth 1
counter-example search depth 2
counter-example search depth 3
counter-example search depth 4
counter-example search depth 5
counter-example search depth 6
counter-example search depth 7
counter-example search depth 8
counter-example search depth 9
counter-example search depth 10
[[1;93mUnknown[0m] Valid up to depth 10
[Lustre] Property safety_missionaries... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Property conservation_cannibals... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Property conservation_missionaries... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Model status: [[1;93mUnknown[0m] 
